extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js

```diff
@@ -1,30 +1,45 @@
-import {Layout, applyFillContent} from '#core/dom/layout';
+import {Layout} from '#core/dom/layout';
 
-const TAG = 'amp-story-shopping-tag';
+import {Services} from '#service';
+
+import {StateProperty} from '../../amp-story/1.0/amp-story-store-service';
 
 export class AmpStoryShoppingTag extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = TAG;
-
-    /** @private {?Element} */
-    this.container_ = null;
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
   }
 
   /** @override */
   buildCallback() {
     super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        this.storeService_ = storeService;
+        this.storeService_.subscribe(
```

Comment:
I don't think you need to subscribe from the `buildCallback` (runs even if the shopping tag is 10 pages away from where the user is), but instead in the `layoutCallback` so we don't use CPU cycles before the element gets close from the viewport

Reply:
Okay, but to be fair I was following the convention in the codelab (which also uses the buildCallback)
The codelab should outline the insight you have provided so I and future contributors to the amp codebase can make better choices during the implementation.